### PR TITLE
fix(typescript-estree, eslint-plugin): stop adding ParenthesizedExpressions to node maps

### PR DESCRIPTION
**Pull request by uniqueiniquity**
Builds off #212 
The core issue for the crashes is that the TS AST has individual nodes for parenthesized expressions, but the ESTree AST does not.
Assuming a tree of the form `(E)`, where `E` is some expression, we would map the TS version of `E` to the ESTree version of `E`, but the ESTree version of `E` would map to the TS version of `(E)`, thereby causing the missing `type` property in the repro.
This PR makes it so we don't add TS nodes of type `ParenthesizedExpression` to the ES->TS map.

I also moved some things around in the particular rule you were investigating; in particular, there's no good reason for us to use the TS node to get the text to check against the provided options, so I switched it to using the ESTree node.

fixes: #187

### CODE REVIEW

File: packages/typescript-estree/src/convert.ts [hunk 1]
```diff
@@ -119,7 +119,12 @@ export class Converter {
 
     if (result && this.options.shouldProvideParserServices) {
       this.tsNodeToESTreeNodeMap.set(node, result);
-      this.esTreeNodeToTSNodeMap.set(result, node);
+      if (node.kind !== SyntaxKind.ParenthesizedExpression) {
```
Comment (armano2) [inline packages/typescript-estree/src/convert.ts]: this is really nice catch, but we should extend it also to include `SyntaxKind.ComputedPropertyName`

its same issue

Comment (armano2) [reply packages/typescript-estree/src/convert.ts]:
```ts
class foo {
	['test']: string
}
```

Comment (armano2) [reply packages/typescript-estree/src/convert.ts]: @JamesHenry if you think this should be addressed in separated pr, i'm ok with that

Comment (uniqueiniquity) [reply packages/typescript-estree/src/convert.ts]: Added the `ComputedPropertyName` case.

**Comment by uniqueiniquity** [comment]
Failing from line ending issues in the fixture

File: packages/eslint-plugin/lib/rules/no-unnecessary-type-assertion.js [hunk 2]
```diff
@@ -162,6 +110,68 @@ module.exports = {
...
   create(context) {
+    const sourceCode = context.getSourceCode();
+
+    /**
+     * @param {Node} node node being linted
+     * @param {Context} context linting context
+     * @param {ts.TypeChecker} checker TypeScript typechecker
+     * @returns {void}
+     */
+    function verifyCast(node, context, checker) {
```
Comment (bradzacher) [inline packages/eslint-plugin/lib/rules/no-unnecessary-type-assertion.js]: remove the `context` parameter.
you can just inherit it from the parent closure.

Comment (uniqueiniquity) [reply packages/eslint-plugin/lib/rules/no-unnecessary-type-assertion.js]: @bradzacher that was really stupid of me, I forgot I had put that there.
In general, do we have a stylistic preference for putting functions inside create and pulling context from the closure vs. putting them outside create and passing context through?

Comment (bradzacher) [reply packages/eslint-plugin/lib/rules/no-unnecessary-type-assertion.js]: Hey man autopilot copy pasta is the best pasta :)

I generally lean toward putting all functions inside create. IMO if it accesses options or context it's better to put it in create so you don't have to pass them around via args.

Comment (uniqueiniquity) [reply packages/eslint-plugin/lib/rules/no-unnecessary-type-assertion.js]: Sounds good. Moved everything into `create`.

**Pull request merged**